apps/gdalalg_vector_pipeline.cpp

```diff
@@ -168,9 +169,29 @@ bool GDALVectorPipelineStepAlgorithm::RunImpl(GDALProgressFunc pfnProgress,
             m_inputDataset.resize(1);
             m_inputDataset[0].Set(readAlg.m_outputDataset.GetDatasetRef());
             m_outputDataset.Set(nullptr);
-            if (RunPreStepPipelineValidations() && RunStep(nullptr, nullptr))
+
+            std::unique_ptr<void, decltype(&GDALDestroyScaledProgress)>
+                pScaledData(nullptr, GDALDestroyScaledProgress);
+
+            const bool bCanHandleNextStep =
+                !bIsStreaming && CanHandleNextStep(&writeAlg);
+
+            if (pfnProgress &&
+                (bCanHandleNextStep || !IsNativelyStreamingCompatible()))
             {
-                if (m_format == "stream")
+                pScaledData.reset(GDALCreateScaledProgress(
+                    0.0, bIsStreaming || bCanHandleNextStep ? 1.0 : 0.5,
+                    pfnProgress, pProgressData));
+            }
+
+            GDALVectorPipelineStepRunContext stepCtxt;
+            stepCtxt.m_pfnProgress = pScaledData ? GDALScaledProgress : nullptr;
+            stepCtxt.m_pProgressData = pScaledData.get();
+            if (bCanHandleNextStep)
+                stepCtxt.m_poNextStep = &writeAlg;
```

Comment:
I see the idea here is to set m_poNextStep only if it can be handled, wouldn't be clearer to always set it and explicitly mark as supported in the context, something like stepCtxt.m_bCanHandleNextStep = true?

Reply:
That seems more complicated to me. It would be little useful for a step to know there is a next step that it cannot handle

Reply:
It's about clarity and readability: if someone reads `if (ctxt.m_poNextStep)` she may think: if it's null the current step must be the last one, right?  Wrong, it's not, it's just that m_poNextStep is not just carrying the information about the next step but what it really also means is that the current step can handle it.

Anyway, not a big issue.

Reply:
maybe rename it as ``m_poNextUsableStep`` ?

Reply:
> maybe rename it as `m_poNextUsableStep` ?

Sounds perfect!